src/plugins/data/common/search/search_source/search_source.ts

```diff
@@ -918,19 +860,188 @@ export class SearchSource {
     // });
     // Alternatively you could also add this query via "Edit as Query DSL", then it needs no code to be changed
 
-    if (highlightAll && body.query) {
-      body.highlight = getHighlightRequest(getConfig(UI_SETTINGS.DOC_HIGHLIGHT));
-      delete searchRequest.highlightAll;
+    body._source = _source;
+
+    // only include unique values
+    if (sourceFieldsProvided && !isEqual(remainingFields, fieldsFromSource)) {
+      setWith(body, '_source.includes', remainingFields, (nsValue) => {
+        return isObject(nsValue) ? {} : nsValue;
+      });
     }
 
-    const omitByIsNil = (object: Record<string, any>) => omitBy(object, isNil);
+    const builtQuery = this.getBuiltEsQuery({
+      index,
+      query: searchRequest.query,
+      filters: searchRequest.filters,
+      getConfig,
+      sort: body.sort,
+    });
 
     const bodyToReturn = {
       ...searchRequest.body,
       pit: searchRequest.pit,
+      query: builtQuery,
+      highlight:
+        searchRequest.highlightAll && builtQuery
+          ? getHighlightRequest(getConfig(UI_SETTINGS.DOC_HIGHLIGHT))
+          : undefined,
+      // remove _source, since everything's coming from fields API, scripted, or stored fields
+      _source: fieldListProvided && !sourceFieldsProvided ? false : body._source,
+      stored_fields:
+        fieldListProvided || sourceFieldsProvided ? [...new Set(remainingFields)] : ['*'],
+      runtime_mappings: runtimeFields,
+      script_fields: scriptedFields,
+      fields: this.getFieldsList({
+        index,
+        fields,
+        docvalueFields: body.docvalue_fields,
+        fieldsFromSource,
+        filteredDocvalueFields,
+        metaFields,
+        fieldListProvided,
+        sourceFieldsProvided,
+      }),
     };
 
-    return omitByIsNil({ ...searchRequest, body: omitByIsNil(bodyToReturn) }) as SearchRequest;
+    return omitByIsNil({
+      ...searchRequest,
+      body: omitByIsNil(bodyToReturn),
+      indexType: this.getIndexType(index),
+      highlightAll:
+        searchRequest.highlightAll && builtQuery ? undefined : searchRequest.highlightAll,
+    }) as SearchRequest;
+  }
+
+  private getFieldFilter({
+    bodySourceExcludes,
+    metaFields,
+  }: {
+    bodySourceExcludes: any;
+    metaFields: any;
+  }) {
+    const filter = fieldWildcardFilter(bodySourceExcludes, metaFields);
+    return (fieldsToFilter: any) =>
+      fieldsToFilter.filter((fld: SearchFieldValue) => filter(this.getFieldName(fld)));
+  }
```

Comment:
>We can also rewrite this function to accept two arguments, instead of an object to avoid another object creation and deconstruction.

What is the downside to using destructured objects as function arguments?

Reply:
Sorry I was too late replying to that: mostly about readability and avoiding creating object that are not needed.
The effort on this PR was to avoid object mutation, cleanup, simplify and this looks to me something that could go hand in hand with that